File: app/src/main/java/es/ulpgc/tfm/ecocsgo/model/Player.java
package es.ulpgc.tfm.ecocsgo.model;

import android.os.Parcel;
import android.os.Parcelable;

import java.util.ArrayList;

public class Player implements Parcelable {
    private ArrayList<MainGunJava> mainGuns;
    private ArrayList<SecondaryGunJava> secondaryGuns;
    private Vest vest;
    private Helmet helmet;
    private DefuseKit defuseKit;
    private boolean alive;

    public Player(EquipmentTeam equipmentTeam){
        mainGuns = new ArrayList<>();
        secondaryGuns = new ArrayList<>();
        alive = true;
    }

    protected Player(Parcel in) {
        mainGuns = in.createTypedArrayList(MainGunJava.CREATOR);
        secondaryGuns = in.createTypedArrayList(SecondaryGunJava.CREATOR);
        vest = in.readParcelable(Vest.class.getClassLoader());
        helmet = in.readParcelable(Helmet.class.getClassLoader());
        defuseKit = in.readParcelable(DefuseKit.class.getClassLoader());
        alive = in.readByte() != 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeTypedList(mainGuns);
        dest.writeTypedList(secondaryGuns);
        dest.writeParcelable(vest, flags);
        dest.writeParcelable(helmet, flags);
        dest.writeParcelable(defuseKit, flags);
        dest.writeByte((byte) (alive ? 1 : 0));
    }

    @Override
    public int describeContents() {
        return 0;
    }

    public static final Creator<Player> CREATOR = new Creator<Player>() {
        @Override
        public Player createFromParcel(Parcel in) {
            return new Player(in);
        }

        @Override
        public Player[] newArray(int size) {
            return new Player[size];
        }
    };

    public ArrayList<MainGunJava> getMainGuns() {
        return mainGuns;
    }

    public ArrayList<SecondaryGunJava> getSecondaryGuns() {
        return secondaryGuns;
    }

    public boolean isAlive() {
        return alive;
    }

    public MainGunJava getLastMainGun(){
        return (mainGuns != null && !mainGuns.isEmpty()) ? mainGuns.get(mainGuns.size() - 1) : null;
    }

    public SecondaryGunJava getLastSecondaryGun(){
        return (secondaryGuns != null && !secondaryGuns.isEmpty()) ? secondaryGuns.get(secondaryGuns.size() - 1) : null;
    }

    public void registerMainGun(MainGunJava gun){
        mainGuns.add(gun);
    }
    public void registerSecondaryGun(SecondaryGunJava gun){
        secondaryGuns.add(gun);
    }

}

File: app/src/main/java/es/ulpgc/tfm/ecocsgo/model/Player.kt
package es.ulpgc.tfm.ecocsgo.model

import android.os.Parcelable
import kotlinx.android.parcel.Parcelize

@Parcelize
data class Player(
    val equipmentTeam: EquipmentTeamEnum,
    var mainWeapons: ArrayList<MainWeapon>? = ArrayList(),
    var secondaryWeapons: ArrayList<SecondaryWeapon>? = ArrayList(),
    var vest: Vest? = null,
    var helmet: Helmet? = null,
    var defuseKit: DefuseKit? = null,
    var alive: Boolean = true
) : Parcelable {

    fun registerMainWeapon(mainWeapon: MainWeapon) {
        updateSelectedWeapon(mainWeapon)
        mainWeapons?.add(mainWeapon)
    }

    fun registerSecondaryWeapon(secondaryWeapon: SecondaryWeapon) {
        updateSelectedWeapon(secondaryWeapon)
        secondaryWeapons?.add(secondaryWeapon)
    }

    fun removeSecondaryWeapon(weaponInGame: Weapon) {
        secondaryWeapons?.remove(weaponInGame)
        if (!secondaryWeapons?.isEmpty()!!) {
            updateSelectedWeapon(secondaryWeapons!!.first())
        }
    }

    fun removeMainWeapon(weaponInGame: Weapon) {
        mainWeapons?.remove(weaponInGame)
        if (!mainWeapons?.isEmpty()!!) {
            updateSelectedWeapon(mainWeapons!!.first())
        }
    }

    fun getMainWeaponInGame(): MainWeapon? {
        for (weapon in mainWeapons!!) {
            if (weapon.inGame) return weapon
        }
        return null
    }

    fun getSecondaryWeaponInGame(): SecondaryWeapon? {
        for (weapon in secondaryWeapons!!) {
            if (weapon.inGame) return weapon
        }
        return null
    }

    fun updateSelectedWeapon(weapon: Weapon) {
        val lastWeaponInGame: Weapon? =
            if (weapon.numeration.category == EquipmentCategoryEnum.PISTOL)
                getSecondaryWeaponInGame() else getMainWeaponInGame()

        if (lastWeaponInGame != null) lastWeaponInGame.inGame = false

        weapon.inGame = true
    }

    fun prepareMainWeaponsForNextRound() {
        val weaponInGame = getMainWeaponInGame()
        mainWeapons?.clear()
        weaponInGame?.casualty = 0
        weaponInGame?.origin = OriginEquipmentEnum.NO_PURCHASED
        weaponInGame?.let { registerMainWeapon(it) }
    }

    fun prepareSecondaryWeaponForNextRound() {
        val weaponInGame = getSecondaryWeaponInGame()
        secondaryWeapons?.clear()
        weaponInGame?.casualty = 0
        weaponInGame?.origin = OriginEquipmentEnum.NO_PURCHASED
        weaponInGame?.let { registerSecondaryWeapon(it) }
    }

    fun prepareUtilityForNextRound() {
        if (vest != null) {
            vest!!.origin = OriginEquipmentEnum.NO_PURCHASED
        }
        if (helmet != null) {
            helmet!!.origin = OriginEquipmentEnum.NO_PURCHASED
        }
        if (defuseKit != null) {
            defuseKit!!.origin = OriginEquipmentEnum.NO_PURCHASED
        }
    }

    fun alreadyHasWeapon(weapon: Weapon): Boolean {
        var hasWeapon = false

        if (weapon.numeration.category == EquipmentCategoryEnum.PISTOL) {
            for (secondaryWeapon in secondaryWeapons!!) {
                if (weapon.name == secondaryWeapon.name) {
                    hasWeapon = true
                    break
                }
            }
        } else {
            for (mainWeapon in mainWeapons!!) {
                if (weapon.name == mainWeapon.name) {
                    hasWeapon = true
                    break
                }
            }
        }

        return hasWeapon
    }

    fun resetPlayer() {
        mainWeapons?.clear()
        secondaryWeapons?.clear()
        resetUtility()
        alive = true
    }

    fun resetUtility() {
        vest = null
        helmet = null
        defuseKit = null
    }
}